File: src/ballbot/ballbot/velocity_controller.py
#!/usr/bin/env python3
from __future__ import annotations

import math
import pigpio
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray, Int32MultiArray

WHEEL_COUNT = 3                                      # liczba kół

WHEEL_STATE_ANGULAR_VEL_IDXS = [4, 9, 14]            # indeksy omega1..omega3 w wheel_state

PIN_RPWM = [10, 4, 6]
PIN_LPWM = [9, 17, 13]
PIN_REN = [11, 8, 19]
PIN_LEN = [5, 22, 26]

PWM_FREQ = 20000
PWM_RANGE = 255

OMEGA_MAX = 330.0 * 2.0 * math.pi / 60.0             # max rad/s
PWM_START_MOVE = 25.0                                # PWM potrzebny do ruszenia
PWM_MAX = 255.0

REF_DEADBAND_OMEGA = 0.05                            # martwa strefa wokół zera
DT_MIN = 1e-3
DT_MAX = 0.05


def clamp(x, lo, hi):
    return max(lo, min(hi, x))                       # ograniczenie zakresu


class PI:
    def __init__(self, kp, ki, i_limit=120.0):
        self.kp = float(kp)                          # wzmocnienie P
        self.ki = float(ki)                          # wzmocnienie I
        self.i_limit = float(i_limit)                # limit całki

        self.integral = 0.0                          # stan całki
        self.initialized = False                     # flaga pierwszego kroku

    def update(self, setpoint, measurement, dt):
        dt = clamp(float(dt), DT_MIN, DT_MAX)       # stabilny krok czasowy

        error = setpoint - measurement               # błąd regulacji e = r - y
        if not self.initialized:
            self.initialized = True                  # inicjalizacja regulatora

        p = self.kp * error                          # człon proporcjonalny

        i_candidate = self.integral + error * dt     # całkowanie błędu
        i_candidate = clamp(i_candidate, -self.i_limit, self.i_limit)  # ograniczenie całki

        u_unsat = p + self.ki * i_candidate          # sygnał przed saturacją

        # Anti-windup przez warunkową akceptację całki
        if u_unsat >= PWM_MAX and error > 0.0:
            pass                                     # nie zwiększaj całki przy dodatnim nasyceniu
        elif u_unsat <= 0.0 and error < 0.0:
            pass                                     # nie zwiększaj całki przy dolnym nasyceniu
        else:
            self.integral = i_candidate              # akceptacja całki, gdy nie pogarsza nasycenia

        return p + self.ki * self.integral, error    # zwracamy też error do logowania

    def reset(self):
        self.integral = 0.0                          # wyzerowanie całki
        self.initialized = False                     # reset stanu


class WheelVelocityMotorNode(Node):
    def __init__(self):
        super().__init__("wheel_velocity_motor_node")  # nazwa noda

        self.ref_sub = self.create_subscription(
            Float64MultiArray,
            "vel_from_controller",
            self.ref_callback,
            10,
        )                                             # referencje prędkości

        self.state_sub = self.create_subscription(
            Float64MultiArray,
            "wheel_state",
            self.state_callback,
            10,
        )                                             # stan kół / prędkości

        self.dir_pub = self.create_publisher(
            Int32MultiArray,
            "motor_direction",
            10,
        )                                             # publikacja kierunku silników

        self.ref_vel = [0.0, 0.0, 0.0]                # zadane prędkości
        self.meas_vel = [0.0, 0.0, 0.0]               # zmierzone prędkości
        self.last_time = self.get_clock().now()       # czas poprzedniej iteracji
        self.last_direction = [0, 0, 0]               # pamięć kierunku do resetu PI

        self.pi_ctrl = [
            PI(kp=10, ki=5, i_limit=120.0),
            PI(kp=10, ki=5, i_limit=120.0),
            PI(kp=10, ki=5, i_limit=120.0),
        ]

        self.pi = pigpio.pi()
        if not self.pi.connected:
            raise RuntimeError("pigpiod not running")

        for i in range(WHEEL_COUNT):
            self.pi.write(PIN_REN[i], 1)
            self.pi.write(PIN_LEN[i], 1)
            self.pi.set_PWM_frequency(PIN_RPWM[i], PWM_FREQ)
            self.pi.set_PWM_frequency(PIN_LPWM[i], PWM_FREQ)
            self.pi.set_PWM_range(PIN_RPWM[i], PWM_RANGE)
            self.pi.set_PWM_range(PIN_LPWM[i], PWM_RANGE)

        self.dir_msg = Int32MultiArray()
        self.dir_msg.data = [0, 0, 0]

        self.log_counter = 0
        self.log_every = 25

        self.stop_all()
        self.get_logger().info("Wheel velocity PI controller started")

    def ref_callback(self, msg):
        if len(msg.data) < WHEEL_COUNT:
            return

        for i in range(WHEEL_COUNT):
            self.ref_vel[i] = float(msg.data[i])

    def apply_motor(self, i, duty, direction):
        duty = int(clamp(duty, 0.0, PWM_MAX))

        if direction > 0:
            self.pi.set_PWM_dutycycle(PIN_LPWM[i], 0)
            self.pi.set_PWM_dutycycle(PIN_RPWM[i], duty)
        elif direction < 0:
            self.pi.set_PWM_dutycycle(PIN_RPWM[i], 0)
            self.pi.set_PWM_dutycycle(PIN_LPWM[i], duty)
        else:
            self.pi.set_PWM_dutycycle(PIN_RPWM[i], 0)
            self.pi.set_PWM_dutycycle(PIN_LPWM[i], 0)

        return duty

    def stop_all(self):
        for i in range(WHEEL_COUNT):
            self.apply_motor(i, 0, 0)
            self.pi_ctrl[i].reset()
            self.last_direction[i] = 0

        self.dir_msg.data = [0, 0, 0]
        self.dir_pub.publish(self.dir_msg)

    def state_callback(self, msg):
        if len(msg.data) < 16:
            return

        for i in range(WHEEL_COUNT):
            self.meas_vel[i] = float(msg.data[WHEEL_STATE_ANGULAR_VEL_IDXS[i]])

        now = self.get_clock().now()
        dt = (now - self.last_time).nanoseconds * 1e-9
        self.last_time = now
        dt = clamp(dt, DT_MIN, DT_MAX)

        dir_out = [0, 0, 0]
        pwm_out = [0.0, 0.0, 0.0]
        err_out = [0.0, 0.0, 0.0]
        int_out = [0.0, 0.0, 0.0]

        for i in range(WHEEL_COUNT):
            ref = self.ref_vel[i]
            meas = self.meas_vel[i]

            if abs(ref) < REF_DEADBAND_OMEGA:
                self.pi_ctrl[i].reset()
                self.apply_motor(i, 0, 0)
                dir_out[i] = 0
                pwm_out[i] = 0.0
                continue

            direction = 1 if ref > 0.0 else -1

            if direction != self.last_direction[i] and self.last_direction[i] != 0:
                self.pi_ctrl[i].reset()

            self.last_direction[i] = direction

            ref_abs = abs(ref)
            meas_abs = abs(meas)

            # CZYSTY PI – brak feed-forward
            u, error = self.pi_ctrl[i].update(ref_abs, meas_abs, dt)

            duty = clamp(u, 0.0, PWM_MAX)

            if duty > 0.0:
                duty = max(PWM_START_MOVE, duty)

            pwm_out[i] = duty
            dir_out[i] = direction
            err_out[i] = error
            int_out[i] = self.pi_ctrl[i].integral

            self.apply_motor(i, duty, direction)

        self.dir_msg.data = dir_out
        self.dir_pub.publish(self.dir_msg)

        self.log_counter += 1
        if self.log_counter >= self.log_every:
            self.log_counter = 0
            self.get_logger().info(
                f"ref={self.ref_vel}, meas={self.meas_vel}, "
                f"err={['{:.2f}'.format(e) for e in err_out]}, "
                f"I={['{:.2f}'.format(i) for i in int_out]}, "
                f"u={['{:.1f}'.format(v) for v in pwm_out]}, "
                f"dir={dir_out}"
            )

    def destroy_node(self):
        self.stop_all()
        self.pi.stop()
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = WheelVelocityMotorNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/ballbot/ballbot/velocity_controller.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <stdexcept>
#include <utility>

#include <pigpiod_if2.h>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "std_msgs/msg/int32_multi_array.hpp"

namespace
{
constexpr size_t WHEEL_COUNT = 3;
constexpr std::array<size_t, WHEEL_COUNT> WHEEL_STATE_ANGULAR_VEL_IDXS{{4, 9, 14}};

constexpr std::array<unsigned, WHEEL_COUNT> PIN_RPWM{{10, 4, 6}};
constexpr std::array<unsigned, WHEEL_COUNT> PIN_LPWM{{9, 17, 13}};
constexpr std::array<unsigned, WHEEL_COUNT> PIN_REN{{11, 8, 19}};
constexpr std::array<unsigned, WHEEL_COUNT> PIN_LEN{{5, 22, 26}};

constexpr unsigned PWM_FREQ = 20000;
constexpr unsigned PWM_RANGE = 255;

constexpr double PWM_START_MOVE = 22.0;
constexpr double PWM_MAX = 255.0;

constexpr double REF_DEADBAND_OMEGA = 0.05;
constexpr double DT_MIN = 1e-3;
constexpr double DT_MAX = 0.05;

double clamp(double x, double lo, double hi)
{
  return std::max(lo, std::min(hi, x));
}

class PIController
{
public:
  PIController(double kp, double ki, double i_limit = 120.0)
  : kp_(kp),
    ki_(ki),
    i_limit_(i_limit)
  {
  }

  std::pair<double, double> update(double setpoint, double measurement, double dt)
  {
    dt = clamp(dt, DT_MIN, DT_MAX);

    const double error = setpoint - measurement;
    if (!initialized_) {
      initialized_ = true;
    }

    integral_ += error * dt;
    integral_ = clamp(integral_, -i_limit_, i_limit_);

    const double p = kp_ * error;
    const double temp = p + ki_ * integral_;

    double u = temp;
    if (temp > PWM_MAX) {
      u = PWM_MAX;
      if (ki_ != 0.0) {
        integral_ = clamp((PWM_MAX - p) / ki_, -i_limit_, i_limit_);
      }
    } else if (temp < 0.0) {
      u = 0.0;
      if (ki_ != 0.0) {
        integral_ = clamp((0.0 - p) / ki_, -i_limit_, i_limit_);
      }
    }

    return {u, error};
  }

  void reset()
  {
    integral_ = 0.0;
    initialized_ = false;
  }

private:
  double kp_;
  double ki_;
  double i_limit_;
  double integral_{0.0};
  bool initialized_{false};
};
}  // namespace

class WheelVelocityMotorNode : public rclcpp::Node
{
public:
  WheelVelocityMotorNode()
  : Node("wheel_velocity_motor_node"),
    last_time_(get_clock()->now())
  {
    ref_sub_ = create_subscription<std_msgs::msg::Float64MultiArray>(
      "vel_from_controller", 1,
      std::bind(&WheelVelocityMotorNode::ref_callback, this, std::placeholders::_1));
    state_sub_ = create_subscription<std_msgs::msg::Float64MultiArray>(
      "wheel_state", 1,
      std::bind(&WheelVelocityMotorNode::state_callback, this, std::placeholders::_1));
    dir_pub_ = create_publisher<std_msgs::msg::Int32MultiArray>("motor_direction", 1);

    pi_ = pigpio_start(nullptr, nullptr);
    if (pi_ < 0) {
      throw std::runtime_error("pigpiod not running");
    }

    for (size_t i = 0; i < WHEEL_COUNT; ++i) {
      gpio_write(pi_, PIN_REN[i], 1);
      gpio_write(pi_, PIN_LEN[i], 1);
      set_PWM_frequency(pi_, PIN_RPWM[i], PWM_FREQ);
      set_PWM_frequency(pi_, PIN_LPWM[i], PWM_FREQ);
      set_PWM_range(pi_, PIN_RPWM[i], PWM_RANGE);
      set_PWM_range(pi_, PIN_LPWM[i], PWM_RANGE);
    }

    dir_msg_.data = {0, 0, 0};
    stop_all();
  }

  ~WheelVelocityMotorNode() override
  {
    cleanup();
  }

private:
  void ref_callback(const std_msgs::msg::Float64MultiArray::SharedPtr msg)
  {
    if (msg->data.size() < WHEEL_COUNT) {
      return;
    }

    for (size_t i = 0; i < WHEEL_COUNT; ++i) {
      ref_vel_[i] = static_cast<double>(msg->data[i]);
    }
  }

  int apply_motor(size_t i, double duty, int direction)
  {
    const int duty_int = static_cast<int>(clamp(duty, 0.0, PWM_MAX));

    if (direction > 0) {
      set_PWM_dutycycle(pi_, PIN_LPWM[i], 0);
      set_PWM_dutycycle(pi_, PIN_RPWM[i], duty_int);
    } else if (direction < 0) {
      set_PWM_dutycycle(pi_, PIN_RPWM[i], 0);
      set_PWM_dutycycle(pi_, PIN_LPWM[i], duty_int);
    } else {
      set_PWM_dutycycle(pi_, PIN_RPWM[i], 0);
      set_PWM_dutycycle(pi_, PIN_LPWM[i], 0);
    }

    return duty_int;
  }

  void stop_all()
  {
    for (size_t i = 0; i < WHEEL_COUNT; ++i) {
      apply_motor(i, 0.0, 0);
      pi_ctrl_[i].reset();
      last_direction_[i] = 0;
    }

    dir_msg_.data = {0, 0, 0};
    dir_pub_->publish(dir_msg_);
  }

  void state_callback(const std_msgs::msg::Float64MultiArray::SharedPtr msg)
  {
    if (msg->data.size() < 16) {
      return;
    }

    for (size_t i = 0; i < WHEEL_COUNT; ++i) {
      meas_vel_[i] = static_cast<double>(msg->data[WHEEL_STATE_ANGULAR_VEL_IDXS[i]]);
    }

    const auto now = get_clock()->now();
    double dt = (now - last_time_).nanoseconds() * 1e-9;
    last_time_ = now;
    dt = clamp(dt, DT_MIN, DT_MAX);

    std::array<int32_t, WHEEL_COUNT> dir_out{{0, 0, 0}};

    for (size_t i = 0; i < WHEEL_COUNT; ++i) {
      const double ref = ref_vel_[i];
      const double meas = meas_vel_[i];

      if (std::abs(ref) < REF_DEADBAND_OMEGA) {
        pi_ctrl_[i].reset();
        apply_motor(i, 0.0, 0);
        dir_out[i] = 0;
        continue;
      }

      const int direction = ref > 0.0 ? 1 : -1;

      if (direction != last_direction_[i] && last_direction_[i] != 0) {
        pi_ctrl_[i].reset();
      }

      last_direction_[i] = direction;

      const double ref_abs = std::abs(ref);
      const double meas_abs = std::abs(meas);

      const auto [u, error] = pi_ctrl_[i].update(ref_abs, meas_abs, dt);
      (void)error;

      double duty = clamp(u, 0.0, PWM_MAX);
      if (duty > 0.0) {
        duty = std::max(PWM_START_MOVE, duty);
      }

      dir_out[i] = direction;
      apply_motor(i, duty, direction);
    }

    dir_msg_.data.assign(dir_out.begin(), dir_out.end());
    dir_pub_->publish(dir_msg_);
  }

  void cleanup()
  {
    if (cleaned_up_) {
      return;
    }
    cleaned_up_ = true;

    if (pi_ >= 0) {
      stop_all();
      pigpio_stop(pi_);
      pi_ = -1;
    }
  }

  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr ref_sub_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr state_sub_;
  rclcpp::Publisher<std_msgs::msg::Int32MultiArray>::SharedPtr dir_pub_;

  std::array<double, WHEEL_COUNT> ref_vel_{{0.0, 0.0, 0.0}};
  std::array<double, WHEEL_COUNT> meas_vel_{{0.0, 0.0, 0.0}};
  rclcpp::Time last_time_;
  std::array<int, WHEEL_COUNT> last_direction_{{0, 0, 0}};
  std::array<PIController, WHEEL_COUNT> pi_ctrl_{{
    PIController(3.5, 23.0, 120.0),
    PIController(3.5, 23.0, 120.0),
    PIController(3.5, 23.0, 120.0)}};

  int pi_{-1};
  bool cleaned_up_{false};
  std_msgs::msg::Int32MultiArray dir_msg_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);

  try {
    auto node = std::make_shared<WheelVelocityMotorNode>();
    rclcpp::spin(node);
  } catch (const std::exception & error) {
    RCLCPP_FATAL(rclcpp::get_logger("velocity_controller"), "%s", error.what());
  }

  if (rclcpp::ok()) {
    rclcpp::shutdown();
  }
  return 0;
}
